waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,145 @@
+module Wasp.LSP.CompletionTest where
+
+import Control.Lens ((^.))
+import Control.Monad.Log.Pure (runLog)
+import Control.Monad.State.Strict (evalStateT)
+import qualified Data.ByteString.Lazy as BS
+import qualified Data.ByteString.Lazy.Char8 as BSC
+import Data.Foldable (find)
+import Data.List (isPrefixOf)
+import Data.Maybe (mapMaybe)
+import qualified Language.LSP.Types as LSP
+import qualified Language.LSP.Types.Lens as LSP
+import System.FilePath (replaceExtension, takeBaseName)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Golden (findByExtension, goldenVsStringDiff)
+import Wasp.Analyzer.Parser.ConcreteParser (parseCST)
+import qualified Wasp.Analyzer.Parser.Lexer as Lexer
+import Wasp.LSP.Completion (getCompletionsAtPosition)
+import Wasp.LSP.ServerState (ServerState (ServerState, _cst, _currentWaspSource, _latestDiagnostics))
+
+-- | Run test cases in ./completionTests directory
+--
+-- See 'readCompletionTest' for the format of a test case.
+test_CompletionLists :: IO TestTree
+test_CompletionLists = do
+  inputFiles <- findByExtension [".wasp"] "./waspls/test/Wasp/LSP/completionTests"
+  return $
+    testGroup "Wasp.LSP.Completion.getCompletionsAtPosition" $
+      makeCompletionTest <$> inputFiles
+
+-- | Takes a completion test case and produces the list of completion items
+runCompletionTest :: String -> String
+runCompletionTest source =
+  let (code, position) = readCompletionTest source
+      tokens = Lexer.lex code
+      parsedCST = snd $ parseCST tokens
+      serverState =
+        ServerState
+          { _currentWaspSource = code,
+            _latestDiagnostics = [],
+            _cst = Just parsedCST
+          }
+      (completionItems, _log) = runLog $ evalStateT (getCompletionsAtPosition position) serverState
+      fmtedCompletionItems = map fmtCompletionItem completionItems
+
+      fmtCompletionItem :: LSP.CompletionItem -> String
+      fmtCompletionItem item =
+        concat
+          [ "  label={",
+            show (item ^. LSP.label),
+            "} kind={",
+            show (item ^. LSP.kind),
+            "} detail={",
+            show (item ^. LSP.detail),
+            "}"
+          ]
```

Comment:
I would also maybe move that prefix of two spaces into line below where you combine them all together, with a `map ("  " <>)`, since that is the place where we care about how they all look together and therefore indentation matters due to the title "Completion Items:\n".

Reply:
I've never actually seen `printf` in Haskell before, cool!